Add sidebar to markersTable (Fixes #915)

zoepage opened the pull request:
First iteration for Marker Table Sidebar


@julienw I have 3 flow errors that do not make sense to me right now. Maybe you can provide more context here?

julienw commented:
I'll try to describe the errors as well as possible so that you understand how flow "sees" the types and so what the errors mean.

Error 1:
```
Error ------------------------------------------------------------------- src/components/sidebar/MarkerSidebar.js:127:17

Cannot create `SidebarDetail` element because in property `value`:
 - Either inexact property `category` of unknown type [1] is incompatible with exact `React.Element` [2].
 - Or property `category` of unknown type [1] is incompatible with `React.Portal` [3].
 - Or property `@@iterator` is missing in property `category` of unknown type [1] but exists in `$Iterable` [4].

   src/components/sidebar/MarkerSidebar.js:127:17
                        v-------------
   127|                 <SidebarDetail
   128|                   label="Category"
   129|                   value={selectedMarker.data.category}
   130|                 />
                        -^

References:
   src/components/sidebar/MarkerSidebar.js:126:16
   126|               {selectedMarker.data.category ? (
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]
   /tmp/flow/flowlib_bd8dd13/react.js:18:5
    18|   | React$Element<any>
            ^^^^^^^^^^^^^^^^^^ [2]
   /tmp/flow/flowlib_bd8dd13/react.js:19:5
    19|   | React$Portal
            ^^^^^^^^^^^^ [3]
   /tmp/flow/flowlib_bd8dd13/react.js:20:5
    20|   | Iterable<React$Node>;
            ^^^^^^^^^^^^^^^^^^^^ [4]
```

The problem here is a type mismatch between the expected type for `value` (`React.Node`, that is [about anything that React can display](https://github.com/facebook/flow/blob/172d28f542f49bbc1e765131c9dfb9e31780f3a2/lib/react.js#L13-L20)) and the type of the passed value `marker.data.category`.

So what is `marker.data.category` ? `marker.data` is a `MarkerPayload` defined in `src/types/markers`, which is an union of different (inexact) types. The fact that the types are inexact is important here. They're inexact because we used `{ ... }` to define them, while exact types use `{| ... |}` or `$Exact<{ ... }>`. "Inexact" means that they must have these properties, **but** they can have others. That's why we try to use exact types as much as possible nowadays, but as you see, we have older code that still use inexact types.

And that's the root of the problem here. While you properly "refined" the type using the ternary, you only checked its existence (and if it's a non-empty string / non zero number of course); but because the types are inexact, it could be present but with an unknown type. That's what the error says.

The easiest way to fix the error is to make the refinement stricter: use `typeof selectedMarker.data.category === 'string'`. Then Flow will know that we pass a string as value, and `string` is a valid `React.Node`.
A less easy way to fix the error is to make all marker payloads exact, but this could uncover other issues. When all payloads are exact, then the mere existence of the `category` means it's a `string`.

Error 2:
```
Error ------------------------------------------------------------------- src/components/sidebar/MarkerSidebar.js:133:16

All branches are incompatible:
 - Either property `interval` is missing in `StyleMarkerPayload_Shared` [1].
 - Or property `interval` is missing in object type [2].

   src/components/sidebar/MarkerSidebar.js:133:16
   133|               {selectedMarker.data.interval ? (
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   src/types/markers.js:367:34
   367| export type StyleMarkerPayload = StyleMarkerPayload_Shared & {
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^ [1]
   src/types/markers.js:367:62
                                                                     v
   367| export type StyleMarkerPayload = StyleMarkerPayload_Shared & {
   368|   cause?: CauseBacktrace,
   369| };
        ^ [2]
```

I'll be honest: I'm not sure why Flow gives this error. But I believe it will be fixed with error 3 below.
Normally it gives this error for unknown properties, but `interval` shouldn't be considered as `unknown` because it's present in other objects of the union.

Error 3:
```
Error ------------------------------------------------------------------- src/components/sidebar/MarkerSidebar.js:134:17

Cannot create `SidebarDetail` element because in property `value`:
 - Either inexact property `interval` of unknown type [1] is incompatible with exact `React.Element` [2].
 - Or property `interval` of unknown type [1] is incompatible with `React.Portal` [3].
 - Or property `@@iterator` is missing in property `interval` of unknown type [1] but exists in `$Iterable` [4].

   src/components/sidebar/MarkerSidebar.js:134:17
                        v-------------
   134|                 <SidebarDetail
   135|                   label="Interval"
   136|                   value={selectedMarker.data.interval}
   137|                 />
                        -^

References:
   src/components/sidebar/MarkerSidebar.js:133:16
   133|               {selectedMarker.data.interval ? (
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]
   /tmp/flow/flowlib_bd8dd13/react.js:18:5
    18|   | React$Element<any>
            ^^^^^^^^^^^^^^^^^^ [2]
   /tmp/flow/flowlib_bd8dd13/react.js:19:5
    19|   | React$Portal
            ^^^^^^^^^^^^ [3]
   /tmp/flow/flowlib_bd8dd13/react.js:20:5
    20|   | Iterable<React$Node>;
            ^^^^^^^^^^^^^^^^^^^^ [4]
```

This is essentially the same error as error 1: if you use `typeof XXX === 'string'` it should work.

julienw commented on src/components/sidebar/index.js:
```diff
@@ -19,7 +20,7 @@ export default function selectSidebar(
     'stack-chart': null,
     'marker-chart': null,
-    'marker-table': null,
+    'marker-table': MarkerSidebar,
```
Note: I don't think we'll want to enable it until it provides new data.

zoepage replied:
This was for testing purpose and will be removed before we merge.

julienw commented on src/components/sidebar/sidebar.css:
```diff
@@ -44,8 +44,9 @@
-  font-size: 12px;
+  font-size: 1.2em;
   margin-bottom: .3em;
+  font-weight: 600;
```
I don't have a strong opinion but I think that in the "Web" context it should be "500" according to https://design.firefox.com/photon/visuals/typography.html.

julienw commented on src/components/sidebar/sidebar.css:
```diff
@@ -44,8 +44,9 @@
 
 .sidebar-title {
-  font-size: 12px;
+  font-size: 1.2em;
```
note that 1.2em is 13.2px. But maybe that's what we want ?

zoepage replied:
We do, yes.

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,157 @@
+  +className?: string,
+|};
+
+class CanSelectContent extends React.PureComponent<CanCopyContentProps> {
```
I think this class could be extracted and shared between the sidebar implementations.

julienw commented:
Here you can find [some more information about refinements from the flow documentation](https://flow.org/en/docs/lang/refinements/).

zoepage commented:
r? @julienw 
I addressed your feedback.

As we talked about reusing [MarkerTooltipContents](https://github.com/devtools-html/perf.html/blob/master/src/components/shared/MarkerTooltipContents.js) for transforming and forwarding the data to the component, I removed the nodes we had in here before as this will be re-done anyway.

julienw commented on src/components/sidebar/index.js:
```diff
@@ -5,12 +5,54 @@
+    );
+  }
+}
+
```
nit: please put this in a separate file instead

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,104 @@
+      <div className="sidebar-value">{value}</div>
+    </React.Fragment>
+  );
+}
```
I think `SidebarDetail` could also be extracted into its own file -- I think this is nice to make it easier to reuse in our various sidebars while still keeping a consistent style.

julienw replied:
Although if we use MarkerTooltipContents in the future we won't likely need SidebarDetail then.

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,104 @@
+              ) : null}
+            </React.Fragment>
+          ) : null}
+        </div>
```
I think you could use MarkerTooltipContents in this PR already. I don't see why waiting for another PR :)

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,104 @@
+type StateProps = {|
+  +selectedNodeIndex: IndexIntoTracingMarkers,
+  +selectedThreadIndex: ThreadIndex,
+  +markers: TracingMarker[],
```
I think that instead of `selectedNodeIndex` and `markers` you could take `marker: TracingMarker | null` directly, that would simplify a bit the code below.

Retrieving the`marker` from the marker index and the `markers` array would be done in  `mapStateToProps`.

I see that the thread index is necessary for `MarkerTooltipContents` though so you can keep it.

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,104 @@
+  render() {
+    const { selectedNodeIndex, markers } = this.props;
+
+    if (selectedNodeIndex === null || selectedNodeIndex === -1) {
```
as a quick note of history, this `-1` value comes from a long time in the project history :) in the future we'll want to use `null`.

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,104 @@
+const options: ExplicitConnectOptions<{||}, StateProps, {||}> = {
+  mapStateToProps: state => ({
+    selectedNodeIndex: selectedThreadSelectors.getViewOptions(state)
+      .selectedMarker,
```
It would be a nice idea to have a `getSelectedMarkerIndex` selector that we could use directly here.

zoepage commented:
After talking to @digitarald, closing in favor of #1442

canaltinova commented:
@julienw Addressed the review comments. Could you look at it again? :)

julienw commented on src/components/sidebar/MarkerSidebar.js:
```diff
@@ -0,0 +1,63 @@
+
+    return (
+      <aside className="sidebar sidebar-calltree">
+        <div className="sidebar-contents-wrapper">
```
The current style for this class isn't very good in our case here (it has a `display: grid` that we don't need here, and makes things look weird). Maybe change the existing `.sidebar-content-wrapper` style to use `.sidebar-calltree .sidebar-content-wrapper` ? Just keep the `line-height` for `.sidebar-content-wrapper`.

canaltinova commented:
Addressed the review comments and rebased :)
